# **Laboratorium 2**


Aleksandra Stachniak, grupa 1

**Łączenie się za pomocą psycopg2 i pandas**

In [2]:
import psycopg2 as pg
import pandas as pd

connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

In [3]:
df = pd.read_sql_query('select * from city',con=connection)
df

city_id                city  country_id         last_update
0          1  A Corua (La Corua)          87 2006-02-15 09:45:25
1          2                Abha          82 2006-02-15 09:45:25
2          3           Abu Dhabi         101 2006-02-15 09:45:25
3          4                Acua          60 2006-02-15 09:45:25
4          5               Adana          97 2006-02-15 09:45:25
..       ...                 ...         ...                 ...
595      596               Zaria          69 2006-02-15 09:45:25
596      597        Zeleznogorsk          80 2006-02-15 09:45:25
597      598         Zhezqazghan          51 2006-02-15 09:45:25
598      599            Zhoushan          23 2006-02-15 09:45:25
599      600          Ziguinchor          83 2006-02-15 09:45:25

[600 rows x 4 columns]

**1. Ile kategorii filmów mamy w wypożyczalni?**

In [4]:
df = pd.read_sql_query('select count(distinct name) from category',con=connection)
df

count
0     16

**Odpowiedź:** W wypożyczalni znajduje się 16 kategorii filmów. Poniżej wyświetlam listę tych kategorii.

In [14]:
df = pd.read_sql_query('select distinct name from category',con=connection)
df

name
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8        Horror
9         Drama
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children

**2. Wyświetl listę kategorii w kolejności alfabetycznej.**

In [20]:
df = pd.read_sql_query('select distinct name from category order by name asc',con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**Odpowiedź:** W poleceniu 1 stosowną listę wyświetliłam, z tym że tym razem należało jeszcze zadbać o kolejność alfabetyczną, co zrobiłam poprzez użycie klauzli *order by* z zastosowaniem znacznika *desc*, co pozwoliło na sortowanie zbioru malejąco.

**3. Znajdź najstarszy i najmłodszy film do wypożyczenia.**

In [7]:
df = pd.read_sql_query('select release_year, title from film',con=connection)
df

release_year              title
0            2006    Chamber Italian
1            2006   Grosse Wonderful
2            2006    Airport Pollock
3            2006  Bright Encounters
4            2006   Academy Dinosaur
..            ...                ...
995          2006     Young Language
996          2006         Youth Kick
997          2006       Zhivago Core
998          2006  Zoolander Fiction
999          2006          Zorro Ark

[1000 rows x 2 columns]

**Odpowiedź:** Po wypisaniu wszystkich tytułów filmów zestawionych z informacją ich wydania można zaobserwować, że dla każdego z nich rok jest ten sam (2006), zatem niemożliwym jest znalezienie najstarszego i najmłodszego filmu do wypożyczenia.

**4. Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?**

In [25]:
df = pd.read_sql_query("select count (rental_date) from rental where rental_date between '2005-07-01' and '2005-08-02'",con=connection)
df

count
0   7380

**Odpowiedź:** W podanym przedziale czasowym odbyło się 7380 wypożyczeń. Poniżej przedstawiam ich konkretne daty.

In [8]:
df = pd.read_sql_query("select rental_date from rental where rental_date between '2005-07-01' and '2005-08-02'",con=connection)
df

rental_date
0    2005-07-05 22:49:24
1    2005-07-05 22:51:44
2    2005-07-05 22:56:33
3    2005-07-05 22:57:34
4    2005-07-05 22:59:53
...                  ...
7375 2005-08-01 23:49:33
7376 2005-08-01 23:50:22
7377 2005-08-01 23:51:00
7378 2005-08-01 23:53:45
7379 2005-08-01 23:58:45

[7380 rows x 1 columns]

**5. Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?**

In [26]:
df = pd.read_sql_query("select count (rental_date) from rental where rental_date between '2010-01-01' and '2011-02-01'",con=connection)
df

count
0      0

**Odpowiedź:** W podanym przedziale czaowym nie został wypożyczony żaden film.

**6. Znajdź największą płatność wypożyczenia.**

In [56]:
df = pd.read_sql_query('select distinct on (amount) amount from payment order by amount desc',con=connection)
df

amount
0    11.99
1    10.99
2     9.99
3     9.98
4     8.99
5     8.97
6     7.99
7     7.98
8     6.99
9     5.99
10    5.98
11    4.99
12    3.99
13    3.98
14    2.99
15    1.99
16    1.98
17    0.99
18    0.00

**Odpowiedź:** Powyżej przedstawam listę wszystkich dokonanych płatności za wypożyczone filmy. Z racji, że interesuje nas największa płatność (11.99), to poniżej sprawdziłam ile ich jest (8) oraz wyszukałam konkretne informacje na temat płatności wypożyczeń o największej wartości.

In [61]:
df = pd.read_sql_query('select count(*) from payment where amount= 11.99',con=connection)
df

count
0      8

In [64]:
df = pd.read_sql_query('select * from payment where amount= 11.99',con=connection)
df

payment_id  customer_id  staff_id  rental_id  amount  \
0       20403          362         1      14759   11.99   
1       22650          204         2      15415   11.99   
2       23757          116         2      14763   11.99   
3       24553          195         2      16040   11.99   
4       24866          237         2      11479   11.99   
5       28799          591         2       4383   11.99   
6       28814          592         1       3973   11.99   
7       29136           13         2       8831   11.99   

                payment_date  
0 2007-03-21 21:57:24.996577  
1 2007-03-22 22:17:22.996577  
2 2007-03-21 22:02:26.996577  
3 2007-03-23 20:47:59.996577  
4 2007-03-02 20:46:39.996577  
5 2007-04-07 19:14:17.996577  
6 2007-04-06 21:26:57.996577  
7 2007-04-29 21:06:07.996577

**7. Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.**

In [71]:
df = pd.read_sql_query("select * from customer_list where country = 'Poland' or country = 'Nigeria' or country = 'Bangladesh'",con=connection)
df

id              name                         address zip code  \
0   284     Sonia Gregory              1279 Udine Parkway    75860   
1   318       Brian Wyman                 1769 Iwaki Lane    25787   
2   572   Sidney Burleson             1405 Hagonoy Avenue    86587   
3   409    Rodney Moeller          1936 Lapu-Lapu Parkway     7122   
4   332    Stephen Qualls                1838 Tabriz Lane     1195   
5   286       Velma Lucas            1155 Liaocheng Place    22650   
6    80      Marilyn Ross                1888 Kabul Drive    20936   
7   198      Elsie Kelley                 955 Bamenda Way     1545   
8   103   Gladys Hamilton                 1177 Jelets Way     3305   
9   329    Frank Waggoner             1816 Bydgoszcz Loop    64308   
10  128   Marjorie Tucker        1050 Garden Grove Avenue     4999   
11   18      Carol Garcia                320 Brest Avenue    43331   
12  270       Leah Curtis                  906 Goinia Way    83565   
13  501       Ruben Geary          414 Mandaluyong Street    16370   
14  277      Olga Jimenez                556 Baybay Manor    55802   
15  156   Bertha Ferguson               1367 Yantai Manor    21294   
16  280    Tracey Barrett  1006 Santa Brbara dOeste Manor    36229   
17  250         Jo Fowler            786 Stara Zagora Way    98332   
18  571  Johnnie Chisholm      1501 Pangkal Pinang Avenue      943   
19  562     Wallace Slone                 1342 Sharja Way    93655   
20   21    Michelle Clark  262 A Corua (La Corua) Parkway    34418   
21  380   Russell Brinson     1642 Charlotte Amalie Drive    75442   
22  558  Jimmie Eggleston            505 Madiun Boulevard    97271   
23  232    Constance Reid               885 Yingkou Manor    31390   

           phone             city     country   notes  sid  
0   195003555232       Benin City     Nigeria  active    1  
1   556100547674        Bydgoszcz      Poland  active    1  
2   867287719310      Czestochowa      Poland  active    1  
3   653436985797        Deba Habe     Nigeria  active    2  
4    38988715447            Dhaka  Bangladesh  active    1  
5   558236142492     Effon-Alaiye     Nigeria  active    1  
6   701457319790              Ife     Nigeria  active    1  
7   768481779568           Ikerre     Nigeria  active    2  
8   484292626944           Ilorin     Nigeria  active    1  
9   965273813662         Jamalpur  Bangladesh  active    2  
10  973047364353  Jastrzebie-Zdrj      Poland  active    1  
11  747791594069           Kaduna     Nigeria  active    2  
12  701767622697           Kalisz      Poland  active    1  
13   52709222667           Lublin      Poland  active    1  
14  363982224739        Ogbomosho     Nigeria  active    2  
15  889538496300             Ondo     Nigeria  active    1  
16   85059738746              Owo     Nigeria  active    2  
17  716256596301              Oyo     Nigeria  active    2  
18  770864062795            Plock      Poland  active    2  
19  946114054231           Sokoto     Nigeria  active    1  
20  892775750063          Tangail  Bangladesh  active    1  
21  821476736117            Tychy      Poland  active    1  
22  970638808606          Wroclaw      Poland  active    1  
23  588964509072            Zaria     Nigeria  active    2

**Odpowiedź:** Powyżej przedstawiam listę wszystkich klientów z Polski, Nigerii oraz Bangladeszu.

**8. Gdzie mieszkają członkowie personelu?**

In [12]:
df = pd.read_sql_query('select distinct country from staff_list',con=connection)
df

country
0  Australia
1     Canada

**Odpowiedź:** Członkowie personelu mieszkają w Australii i Kanadzie.

**9. Ilu pracowników mieszka w Argentynie lub Hiszpanii?**

In [13]:
df = pd.read_sql_query("select count(*) from staff_list where country = 'Argentina' or country = 'Spain'",con=connection)
df

count
0      0

**Odpowiedź:** Już na podstawie 8 zadania można wywnioskować, że nikt nie mieszka w Argentynie lub Hiszpanii. Potwierdziała to również odpowiednio sformułowana instrukcja z zapytaniem o liczbę pracowników mieszkających w danych państwach.

**10. Jakie kategorie filmów zostały wypożyczone przez klientów?**

In [116]:
df = pd.read_sql_query("select distinct category.name from rental\n" 
" left join inventory on rental.inventory_id = inventory.inventory_id\n" 
" left join film on film.film_id = inventory.film_id\n" 
" left join film_category on film_category.film_id = film.film_id\n" 
" left join category on film_category.category_id = category.category_id",con=connection)
df

name
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8         Drama
9        Horror
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children

**Odpowiedź:** Powyżej znajduje się lista wszystkitch kategorii filmów wypożyczonych przez klientów.

**11. Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.**

In [111]:
df = pd.read_sql_query("select distinct on (category.name) name from rental\n" 
" left join inventory on rental.inventory_id = inventory.inventory_id\n" 
" left join film on film.film_id = inventory.film_id\n" 
" left join film_category on film_category.film_id = film.film_id\n" 
" left join category on film_category.category_id = category.category_id\n" 
" left join customer on customer.customer_id = rental.customer_id\n" 
" left join address on address.address_id = customer.address_id\n" 
" left join city on city.city_id = address.city_id\n" 
" left join country on country.country_id = city.country_id where country.country='United States'",con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

**Odpowiedź:** Powyżej znajduje się lista wszystkich kategorii filmów wypożyczonych w Ameryce.

**12. Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley.**

In [105]:
df = pd.read_sql("select title, actor.first_name, actor.last_name from film\n" 
" left join film_actor on film_actor.film_id = film.film_id\n" 
" left join actor on actor.actor_id = film_actor.actor_id\n" 
" where (actor.first_name='Olympia' and actor.last_name='Pfeiffer') or (actor.first_name='Julia' and actor.last_name='Zellweger')\n" 
" or (actor.first_name='Ellen' and actor.last_name='Presley')",con=connection)
df

title first_name  last_name
0     Bilko Anonymous      Ellen    Presley
1   Caribbean Liberty      Ellen    Presley
2    Casper Dragonfly      Ellen    Presley
3    Empire Malkovich      Ellen    Presley
4       Floats Garden      Ellen    Presley
..                ...        ...        ...
64       Outlaw Hanky      Julia  Zellweger
65    Panky Submarine      Julia  Zellweger
66   Rider Caddyshack      Julia  Zellweger
67          Won Dares      Julia  Zellweger
68      Wyoming Storm      Julia  Zellweger

[69 rows x 3 columns]

**Odpowiedź:** Powyżej znajduje się lista filmów z aktorami, o których mowa w poleceniu.